### PR TITLE
Minimised Input textbox labelling

### PR DESCRIPTION
**Pull request by rossbulat**
Aiming to solve the following issues in regards to text input with this PR:

* Alignment: Text input labels have arbitrary widths and do not align in most cases
* Mobile friendly: Smaller screens do not display labels beside a textbox well
* Space: Normalises spacing, more coherent when multiple inputs are grouped in rows

--
Embedded in `<Labelled />`  is an added functionality to embed a textbox label inside the field, and to minimise it outwards when the field has a value or is focussed:

textbox1

The textbox is smart enough not to do this when there is no label:

Screenshot 2019-04-08 at 00 10 39

Labels are minimised by default if a `defaultValue` is provided:

Screenshot 2019-04-08 at 00 12 35

And placeholders are maintained when a field is focussed:

Screenshot 2019-04-08 at 00 13 44

And `<Labelled />` is still fully intact when not accompanied by a text input:

Screenshot 2019-04-08 at 00 14 05

### CODE REVIEW

**Comment by rossbulat** [comment]
Working on resolving conflicts.

**Comment by jacogr** [comment]
I would handle this the other way around -

- keep as it for larger displays
- move to on-top as the display is small (originally it was there, however it actually leads to a lot of wasted space on larger width screens)

For something like transfer, moving the things up and down leads to some jumping otherwise. Additionally, there are some fields which will never have placeholders, i.e. selections for an account. So you will end up with some above, some internal and some others above when moved into.

**Comment by rossbulat** [comment]
Agreed that spacing is quite generous, I have decreased slightly in my latest commit.

Rather than condense space as much as possible, I would go in the direction of utilising as much of the screen estate at our disposal as we can in any given app. Right now this is not the case, even in the event where they is lots of space to spare. My thoughts here are to optimise the spacing for friendly, predictable UI,  and work towards a scenario where:

a.  There are 2 - maybe 3 max - inputs on screen at one time. 4 - 5 is too many.

b. Utilise available screen estate as much as possible; I'd actually like to see the inputs larger than they are now, which will be more viable if point (a) is fulfilled. 

> keep as it for larger displays

Well I would prioritise consistency and readability over space savings at this point. Right now none of the forms (on most display sizes >= iPad mini) take up more than ~50% of height. Let's take a look at the `Create Account` form  (the longest form in the app right now minus the toolbox?):


Screenshot 2019-04-08 at 01 33 43
Above labelling before, side-by-side labelling after.

Quite a bit of free space there in both cases, with the above labelling taking ~77px more space, with still a lot to spare.

> For something like transfer, moving the things up and down leads to some jumping otherwise. 

The content will not be jumping as `<StyledLabelled />`'s padding is consistent throughout states.

> Additionally, there are some fields which will never have placeholders,

So in this PR placeholders show up once you are in focus - the label of the input is primarily displayed internally. The design is friendly for placeholder and non-placeholder inputs.

I'd also argue that dedicating a full row to labels will give more space for help prompts. Having label + help prompt + input field all in one row may be space limiting and confusing for the user.

To summarise
- there is still lots of height to utilise with forms, even on desktops with 5 fields
- would like to see form elements being simplified to support 2 / 3 inputs per screen, allowing more room for validating and ad hoc elements.
- padding is maintained in all states of the element so there is no jumping around

Perhaps a personal preference of mine, but I would also like to further minimise the input box itself and animate an underlying bar as the input is being focussed / blurred. This design would compliment an internal label very well too.

**Comment by jacogr** [comment]
So the example above is mixed - some stuff has side-by-side, eg. the crypto doesn't have the labelling. Additionally, it looks off since the password label is not there. I'm also worrried about the cases where we actually have the help - now it is hidden by default and visible only when something is changed.

As for 4-5 inputs. There are 2 communities active -

- normal users which are using apps available on their chains, e.g. contracts for substrate, democrcy (when the chain supports it), transfer, etc.
- developers and chain builders with completely custom runtimes, here they live inside the extrinsics app (mostly). These I'm not sure works at all - even doing a simple transfer (we had this originally), with labels on-top just makes it all unwieldly.

So would come back to -

- status quo for normal screens
- labels on top for small ones (here we don't need to worry about any "special" usage)

**Comment by rossbulat** [comment]
Indeed, there are a number of formatting issues encapsulated here and the proposed labelling does not fit all use cases at this time.

To address, my thoughts are to start with a mobile first approach to coincide with the min-width styling rules. Would you support this plan of action?:-

1. Default labels to above placement for textboxes and dropdowns throughout the app, (no animation)

2. Introduce optimised desktop side-by-side label layout via #902 setup

3. Optimise styling - perhaps some focus / blur visual cues for each theme.

4. For edge cases, a screen width media query will not alway suffice to determine the label placement. E.g. a containing element's width may be constrained even on desktop screen size. In this case we could adopt a [component measuring](https://www.npmjs.com/package/react-measure) HOC to determine the labelling placement.

This would give us the granularity of determining the best label placement for a components that make our lives more difficult.

5 (optional). For minimal enlarged inputs (admittedly not really an opportunity to do so at this time), adopt the label animation that I proposed in this PR.

6. Merge 🎉

**Comment by jacogr** [comment]
Agreed. Baby steps. Let's solve the small issue first and then can look at building on that. (So on-board with as-is and above for smaller as the first step)
